server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,744 @@
+            this.docStartFP = docStartFP;
+            this.posStartFP = posStartFP;
+            this.payloadFP = payloadFP;
+        }
+
+        long distance(BlockTermState other) {
+            return this.docStartFP - other.docStartFP + this.posStartFP - other.posStartFP + this.payloadFP - other.payloadFP;
+        }
+    }
+
+    void analyzePostings(SegmentReader reader, IndexDiskUsageStats stats) throws IOException {
+        FieldsProducer postingsReader = reader.getPostingsReader();
+        if (postingsReader == null) {
+            return;
+        }
+        postingsReader = postingsReader.getMergeInstance();
+        PostingsEnum postings = null;
+        for (FieldInfo field : reader.getFieldInfos()) {
+            if (field.getIndexOptions() == IndexOptions.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetBytesRead();
+            final Terms terms = postingsReader.terms(field.name);
+            if (terms == null) {
+                continue;
+            }
+            // It's expensive to look up every term and visit every document of the postings lists of all terms.
+            // As we track the min/max positions of read bytes, we just visit the two ends of a partition containing
+            // the data. We might miss some small parts of the data, but it's an good trade-off to speed up the process.
+            TermsEnum termsEnum = terms.iterator();
+            if (terms instanceof FieldReader) {
+                final BlockTermState minState = Objects.requireNonNull(
+                    getBlockTermState(termsEnum, terms.getMin()), "can't retrieve the block term state of the min term");
+                final BlockTermState maxState = Objects.requireNonNull(
+                    getBlockTermState(termsEnum, terms.getMax()), "can't retrieve the block term state of the max term");
+                final long skippedBytes = maxState.distance(minState);
+                stats.addInvertedIndex(field.name, skippedBytes);
+                termsEnum.seekExact(terms.getMax());
+                postings = termsEnum.postings(postings, PostingsEnum.ALL);
+                if (postings.advance(termsEnum.docFreq() - 1) != DocIdSetIterator.NO_MORE_DOCS) {
+                    readProximity(terms, postings);
+                }
+                final long bytesRead = directory.getBytesRead();
+                int visitedTerms = 0;
+                final long totalTerms = terms.size();
+                termsEnum = terms.iterator();
+                // Iterate until we really access the first terms, but iterate all if the number of terms is small
+                while (termsEnum.next() != null) {
+                    cancellationChecker.logEvent();
+                    ++visitedTerms;
+                    if (totalTerms > 1000 & visitedTerms % 50 == 0 && directory.getBytesRead() > bytesRead) {
```

Comment:
Should this be `&&` and not `&`? Only the former is lazy.

Reply:
Thanks for catching this :). I fixed it in https://github.com/elastic/elasticsearch/pull/74051/commits/6083c1416781bae73b4d7fd08ebdb23c6179ce04.